core/src/main/java/io/dekorate/utils/TopologicalSort.java

```diff
@@ -0,0 +1,225 @@
+import io.dekorate.kubernetes.decorator.Decorator;
+
+public final class TopologicalSort {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TopologicalSort.class);
+
+  private TopologicalSort() {
+
+  }
+
+  /**
+   * Adapt the order driven by before and after dependencies into a direct direction `dependsOn`.
+   * 
+   * @param items
+   * @return
+   */
+  public static List<Decorator> sortDecorators(Collection<Decorator> items) {
+    // This dictionary is used to verify a decorator is in place and register the hierarchy.
+    Map<Class, List<Decorator>> dictionary = createDictionary(items);
+    Map<Object, Node<Decorator>> unordered = new HashMap<>();
+    for (Decorator item : items) {
+      Class identity = item.getClass();
+      Class[] after = item.after();
+      Class[] before = item.before();
+      Node<Decorator> node = unordered.get(identity);
+      if (node == null) {
+        node = new Node<>(identity);
+        unordered.put(identity, node);
+      }
+
+      node.references.add(item);
+      if (after != null) {
+        // the after order is the same as the dependsOn order, so we simply add it.
+        for (Class a : after) {
+          List<Decorator> afterDecorators = dictionary.get(a);
+          if (afterDecorators != null) {
+            for (Decorator afterDecorator : afterDecorators) {
+              if (!afterDecorator.getClass().equals(identity)) {
+                node.getDepends().add(afterDecorator.getClass());
+              }
+            }
+          } else if (isVerbose()) {
+            LOGGER.info("[sort] Warning. Declared decorator in " + identity + " is not found " + a);
+          }
+        }
+      }
+
+      if (before != null) {
+        // for before order, we need to find the nodes that represent it and add the current node to its dependsOn field.
+        for (Class b : before) {
+          List<Decorator> beforeDecorators = dictionary.get(b);
+          if (beforeDecorators != null) {
+            for (Decorator beforeDecorator : beforeDecorators) {
+              if (!beforeDecorator.getClass().equals(identity)) {
+                Node<Decorator> beforeNode = unordered.get(beforeDecorator.getClass());
+                if (beforeNode == null) {
+                  beforeNode = new Node<>(beforeDecorator.getClass());
+                  unordered.put(beforeDecorator.getClass(), beforeNode);
+                }
+
+                beforeNode.getDepends().add(identity);
+                // to avoid
+                node.getDepends().remove(beforeDecorator.getClass());
+              }
+            }
+          } else if (isVerbose()) {
+            LOGGER.info("[sort] Warning. Declared decorator in " + identity + " is not found " + b);
+          }
+        }
+      }
+    }
+
+    return sort(unordered.values());
+  }
+
+  /**
+   * Sort nodes with dependencies using a Graph Topological algorithm plus the visitor pattern to avoid having cycles and
```

Comment:
Can your provide more details about the algorithm used ?

Reply:
Yes, I'm planning to refactor and clean the current code.